pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -3061,39 +3062,67 @@ public CompletableFuture<Void> onPoliciesUpdate(@Nonnull Policies data) {
-                return CompletableFuture.allOf(replicationFuture, dedupFuture, persistentPoliciesFuture,
-                        preCreateSubscriptionForCompactionIfNeeded());
-            });
-        }).exceptionally(ex -> {
-            log.error("[{}] update namespace polices : {} error", this.getName(), data, ex);
-            throw FutureUtil.wrapToCompletionException(ex);
+        // Client permission check.
+        subscriptions.forEach((subName, sub) -> {
+            sub.getConsumers().forEach(consumer -> applyPoliciesFutureList.add(consumer.checkPermissionsAsync()));
         });
+        producers.values().forEach(producer -> applyPoliciesFutureList.add(
+                producer.checkPermissionsAsync().thenRun(producer::checkEncryption)));
+        // Check message expiry.
+        applyPoliciesFutureList.add(CompletableFuture.runAsync(
+                () -> checkMessageExpiry(), MoreExecutors.directExecutor()));
+
+        // Update rate limiters.
+        applyPoliciesFutureList.add(CompletableFuture.runAsync(
+                () -> updateDispatchRateLimiter(), MoreExecutors.directExecutor()));
+        applyPoliciesFutureList.add(CompletableFuture.runAsync(
+                () -> updateSubscribeRateLimiter(), MoreExecutors.directExecutor()));
+        applyPoliciesFutureList.add(CompletableFuture.runAsync(
+                () -> updatePublishRateLimiter(), MoreExecutors.directExecutor()));
+        if (namespaceLevelPolicies != null) {
+            applyPoliciesFutureList.add(CompletableFuture.runAsync(
+                    () -> updateResourceGroupLimiter(namespaceLevelPolicies), MoreExecutors.directExecutor()));
+        }
+        applyPoliciesFutureList.add(CompletableFuture.runAsync(
+                () -> updateSubscriptionsDispatcherRateLimiter(), MoreExecutors.directExecutor()));
+        applyPoliciesFutureList.add(CompletableFuture.runAsync(
+                () -> replicators.forEach((name, replicator) -> replicator.updateRateLimiter()),
+                MoreExecutors.directExecutor()));
+        applyPoliciesFutureList.add(CompletableFuture.runAsync(
+                () -> shadowReplicators.forEach((name, replicator) -> replicator.updateRateLimiter()),
+                MoreExecutors.directExecutor()));
+
+        // Other components.
+        applyPoliciesFutureList.add(CompletableFuture.runAsync(
+                () -> checkReplicationAndRetryOnFailure(), MoreExecutors.directExecutor()));
+        applyPoliciesFutureList.add(CompletableFuture.runAsync(
+                () -> checkDeduplicationStatus(), MoreExecutors.directExecutor()));
+        applyPoliciesFutureList.add(CompletableFuture.runAsync(
+                () -> checkPersistencePolicies(), MoreExecutors.directExecutor()));
+        applyPoliciesFutureList.add(CompletableFuture.runAsync(
+                () -> preCreateSubscriptionForCompactionIfNeeded(), MoreExecutors.directExecutor()));
+
+        return applyPoliciesFutureList;
```

Comment:
Add a new private method :
```
private CompletableFuture<Void> runDirectly(Runnable runnable) {
        return CompletableFuture.runAsync(runnable, MoreExecutors.directExecutor());
    }
```

Reply:
Sure, added a tool method in `FutureUtil`